ssl/statem/statem_srvr.c

```diff
@@ -114,6 +114,17 @@ int ossl_statem_server_read_transition(SSL *s, int mt)
                 return 1;
             }
         }
+        if (mt == SSL3_MT_CLIENT_KEY_EXCHANGE && s->s3->tmp.cert_request
```

Comment:
Yes, your `if` structure is better. However, I'm not convinced about the separate `client_certificate_expected` method idea. The problem is that, to actually make things simpler, you would also have to offload the "send a handshake failure alert instead of an unexpected message" processing to that function too. I'm not sure that is a good idea.